atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.common.compression;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+
+import com.google.common.base.Preconditions;
+
+import net.jpountz.lz4.LZ4SafeDecompressor;
+import net.jpountz.xxhash.StreamingXXHash32;
+import net.jpountz.xxhash.XXHashFactory;
+
+/**
+ * {@link InputStream} that wraps an InputStream that is compressed
+ * according to the v1.5.0 LZ4 specification. Buffers and decompresses
+ * the content of the delegate stream.
+ */
+public class LZ4DecompressingInputStream extends BufferedDelegateInputStream {
+
+    private final LZ4SafeDecompressor decompressor;
+    private final LZ4FrameDescriptor frameDescriptor;
+    private final byte[] compressedBuffer;
+    private final StreamingXXHash32 hasher;
+
+    // Length in bytes of the next compressed block to be read from the delegate
+    private int nextLength;
+
+    public LZ4DecompressingInputStream(InputStream delegate) throws IOException {
+        super(delegate);
+        this.decompressor = LZ4Streams.decompressor;
+        this.hasher = XXHashFactory.fastestInstance().newStreamingHash32(0);
+        this.frameDescriptor = readFrameHeader();
+        this.compressedBuffer = new byte[LZ4Streams.getCompressedBufferSize(frameDescriptor)];
+
+        allocateBuffer(LZ4Streams.getUncompressedBufferSize(frameDescriptor));
+    }
+
+    // Reads the frame header as well as the size of the first block
+    private LZ4FrameDescriptor readFrameHeader() throws IOException {
+        byte[] frameHeader =
+                new byte[LZ4Streams.MAGIC_LENGTH + LZ4Streams.FRAME_DESCRIPTOR_LENGTH
+                         + LZ4Streams.BLOCK_HEADER_LENGTH];
+        nextLength = readBlock(frameHeader, frameHeader.length);
+        Preconditions.checkState(LZ4Streams.intFromLittleEndianBytes(frameHeader, 0) == LZ4Streams.MAGIC_VALUE,
+                "Input is not an lz4 stream that this input stream can decompress");
+        return LZ4FrameDescriptor.fromByteArray(Arrays.copyOfRange(frameHeader,
+                LZ4Streams.MAGIC_LENGTH, LZ4Streams.MAGIC_LENGTH + LZ4Streams.FRAME_DESCRIPTOR_LENGTH));
+    }
+
+    // This reads one block plus the size of the next block in as few
+    // read calls as possible. The length should be the number of bytes
+    // of data plus HEADER_SIZE bytes for the length of the next block.
+    private int readBlock(byte[] b, int length) throws IOException {
+        int bytesReadSoFar = 0;
+        while (bytesReadSoFar < length) {
+            int numRead = delegate.read(b, bytesReadSoFar, length - bytesReadSoFar);
```

Comment:
I think the docs on InputStream suggest that this shouldn't happen. https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#read(byte[],%20int,%20int) indicates that the read call should block until bytes are available. The default, inefficient, implementation blocks until the requested number of bytes is read or end of files is detected.

Alternatively, we could read from the delegate until a -1 is explicitly received, but I don't think this is necessary.